libraries/resource_provider.rb

```diff
@@ -19,35 +19,67 @@ class ResourceProvider
+          else
+            OneviewCookbook::Helper.build_client(context.client)
+          end
+      new_data = JSON.parse(context.data.to_json) rescue context.data
+      @item = context.property_is_set?(:api_header_version) ? klass.new(c, new_data, context.api_header_version) : klass.new(c, new_data)
+      @item['name'] ||= context.name
+    end
+
+    # Helper method that analyzes the namespace and defines the class variables and the resource class itself
+    # @return [OneviewSDK::Resource] SDK resource class
+    # @raise [NameError] If for some reason the method could not resolve the namespace to a OneviewSDK::Resource
+    def parse_namespace
+      klass_name = self.class.to_s
+      Chef::Log.debug("Resource '#{@resource_name}' received with the '#{klass_name}' class name")
+      name_arr = klass_name.split('::')
+      @sdk_resource_type = name_arr.pop.gsub(/Provider$/i, '') # e.g., EthernetNetwork
       case name_arr.size
-      when 2 # No variant or api version specified. (OneviewCookbook::ResourceProvider)
+      when 1
+        Chef::Log.debug("Resource '#{klass_name}' has no variant or api version specified.")
         # This case should really only be used for testing
         # Uses the SDK's default api version and variant
-        @sdk_resource_type = name_arr.pop.gsub(/Provider$/i, '') # e.g., EthernetNetwork
-        klass = OneviewSDK.resource_named(@sdk_resource_type)
-      when 3 # No variant specified. e.g., OneviewCookbook::API200::EthernetNetworkProvider
-        @sdk_resource_type = name_arr.pop.gsub(/Provider$/i, '') # e.g., EthernetNetwork
+        return OneviewSDK.resource_named(@sdk_resource_type)
+      when 2 # No variant specified. e.g., OneviewCookbook::API200::EthernetNetworkProvider
+        Chef::Log.debug("Resource '#{klass_name}' has no variant specified.")
         @sdk_api_version = name_arr.pop.gsub(/API/i, '').to_i # e.g., 200
         @sdk_variant = nil # Not needed
-      when 4 # The variant is specified. e.g., OneviewCookbook::API200::C7000::EthernetNetworkProvider
-        @sdk_resource_type = name_arr.pop.gsub(/Provider$/i, '') # e.g., EthernetNetwork
+      when 3
+        parse_namespace_with_secondary_apis(klass_name, name_arr)
+      else # Something is wrong
+        raise NameError, "Can't build a resource object from the class #{self.class}"
+      end
+      base_sdk = @sdk_secondary_api ? OneviewSDK.const_get(@sdk_secondary_api) : OneviewSDK
+      klass ||= base_sdk.resource_named(@sdk_resource_type, @sdk_api_version, @sdk_variant)
+      klass
+    end
+
+    # Parses all the namespace whrn it is fully specified with API version, variant or secondary APIs
+    # @param [String] klass_name Name of the provider class
+    # @param [Array<String>] name_arr Remaining namespace list that may contain variant, api version and secondary api name
+    def parse_namespace_with_secondary_apis(klass_name, name_arr)
```

Comment:
I think all this should be included in the method above. If we need to disable the method-size rubocop rule with comments around it that's fine, but splitting it is confusing.